src/Network/NWEndpoint.cs

```diff
@@ -0,0 +1,163 @@
+			var handle = nw_endpoint_create_host (hostname, port);
+			if (handle == IntPtr.Zero)
+				return null;
+			return new NWEndpoint (handle, owns: true);
+		}
+
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern string nw_endpoint_copy_port_string (OS_nw_endpoint endpoint);
+
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		public string Port => nw_endpoint_copy_port_string (handle);
+
+		
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern ushort nw_endpoint_get_port (OS_nw_endpoint endpoint);
+
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		public ushort PortNumber => nw_endpoint_get_port (handle);
+	
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern OS_nw_endpoint nw_endpoint_create_address (IntPtr sockaddrPtr);
+
+		// TODO: .NET has a SocketAddress type, we could use it for simplicity, but the
+		// address family would have to be mapped, and it does not look like a very useful
+		// type to begin with.
+
+		// extern char * _Nonnull nw_endpoint_copy_address_string (nw_endpoint_t _Nonnull endpoint) __attribute__((availability(tvos, introduced=12.0))) __attribute__((availability(watchos, introduced=5.0))) __attribute__((availability(ios, introduced=12.0))) __attribute__((availability(macos, introduced=10.14)));
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern string nw_endpoint_copy_address_string (OS_nw_endpoint endpoint);
+		
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		public string Address => nw_endpoint_copy_address_string (handle);
+		
+		// extern const struct sockaddr * _Nonnull nw_endpoint_get_address (nw_endpoint_t _Nonnull endpoint) __attribute__((availability(tvos, introduced=12.0))) __attribute__((availability(watchos, introduced=5.0))) __attribute__((availability(ios, introduced=12.0))) __attribute__((availability(macos, introduced=10.14)));
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern IntPtr /* struct sockaddr* */ nw_endpoint_get_address (OS_nw_endpoint endpoint);
+
+		// TODO: same
+	
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern unsafe OS_nw_endpoint nw_endpoint_create_bonjour_service (string name, string type, string domain);
+
+		public NWEndpoint CreateBonjourService (string name, string serviceType, string domain)
+		{
+			if (serviceType == null)
+				throw new ArgumentNullException (nameof (serviceType));
+			var x = nw_endpoint_create_bonjour_service (name, serviceType, domain);
+			return new NWEndpoint (x, owns: true);
+		}
+		
+	
+		[Watch (5,0), TV (12,0), Mac (10,14), iOS (12,0)]
+		[DllImport (Constants.NetworkLibrary)]
+		static extern unsafe string nw_endpoint_get_bonjour_service_name (OS_nw_endpoint endpoint);
```

Comment:
This P/Invoke will free the string returned from `nw_endpoint_get_bonjour_service_name`, but it's defined as `const char *` in the header, so it shouldn't be freed.

Reply:
good catch!